api/src/main/java/org/apache/iceberg/expressions/MaxAggregate.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.expressions;
+
+import java.util.Comparator;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.types.Comparators;
+import org.apache.iceberg.types.Conversions;
+import org.apache.iceberg.types.Type.PrimitiveType;
+import org.apache.iceberg.types.Types;
+
+public class MaxAggregate<T> extends ValueAggregate<T> {
+  private final int fieldId;
+  private final PrimitiveType type;
+  private final Comparator<T> comparator;
+
+  protected MaxAggregate(BoundTerm<T> term) {
+    super(Operation.MAX, term);
+    Types.NestedField field = term.ref().field();
+    this.fieldId = field.fieldId();
+    this.type = field.type().asPrimitiveType();
+    this.comparator = Comparators.forType(type);
+  }
+
+  @Override
+  protected boolean hasValue(DataFile file) {
+    return file.valueCounts().containsKey(fieldId);
```

Comment:
For the value to be valid, it must be expected to be non-null and should actually be included in upper bounds.

That is:
* If the column is all null values (`safeGet(file.valueCounts(), fieldId) != null && safeGet(file.valueCounts(), fieldId) == safeGet(file.nullValueCounts(), fieldId)`) then this returns `true` because `null` is valid
* If the column has at least one non-null value, then `file.upperBounds()` must contain `fieldId`.

Reply:
Fixed. Thanks!